Proposal for formalizing review process for PRs

KevinOConnor opened the pull request:
I'm hopelessly behind on reviews of Klipper PRs, and the unfortunate reality is that I do not think I will be able to get through the backlog.  This PR is intended to start a discussion to see if we can get more people to help perform reviews.

I find that the majority of feedback I give on reviews are for "procedural items".  I find that I frequently make similar comments (eg, must have signed-off-by line, must follow documentation, must update documentation) on many reviews.  I think this work done to facilitate merging (while trying to keep a high-quality submission process) could be spread to many more people.

The proposed process is documented at https://github.com/Klipper3d/klipper/blob/work-review-20211116/docs/CONTRIBUTING.md .

This is a proposal and I'm looking for feedback on it.  I can write all I want on this topic, but it wont mean anything if there are no reviewers willing to volunteer.  Thus, I view this as an "open discussion".

If there is interest in being a reviewer, then please read the document above and let me know your thoughts on it as well as the areas you are interested in reviewing.  In general, I expect "reviewers" to have had an ongoing history working with users on Klipper Discord, Klipper Discourse, or working on PRs on Klipper GitHub.  Being a software developer is not a prerequisite.

@dmbutyugin , @Arksine - I added you to the list of reviewers in the document above as an "example" - don't feel compelled to take on those roles.

Thoughts?
-Kevin

mental405 commented:
The request looks good. It pretty concisely lays out the base requirements for a PR as well as setting some expectations for the submitter.

You can put me down for "configuration files and documentation" review if you would like.

Sineos commented:
Unfortunately, I cannot contribute to any code related items but for the formal stuff, I'd be willing to step in

JamesH1978 commented:
Very concise doc, I would be willing to look at config files and documentation

grigi commented on docs/CONTRIBUTING.md:
```diff
@@ -1,18 +1,238 @@
+| Dmitry Butyugin        | @dmbutyugin       | Input shaping, resonance testing, kinematics |
+| Kevin O'Connor         | @KevinOConnor     | Core motion system, Micro-controller code |
+
+Please do not "ping" any of the reviewers and please do not direct
+submissions at them. All of the reviewers monitor the forums and PRs,
+and will take on reviews when they have time to.
```
If we have been working on a PR, and once its deemed complete, how do we notify anyone that it's worth their time to actually pay attention?

Adding a tag/label would be about the right level of visibility, but Github doesn't have enough permission granularity that would allow the random person to set a few tags, but not others.

Starting all pr's in draft state could help a bit, but I think it means different enough things to different people that it isn't a solution in itself.

eduncan911 replied:
The answer is to use GitHub Actions to auto-tag a PR once it is done.

There are many ways to skin this cat; but, having a way that an author can make their changes as "Now ready for Review", usually in a comment with a special keyword, works best.

The overall process works best by auto-warning that their PR is about to be closed, and then auto-closing after a set time (30 days, 90 days, etc).

That way, the author is tagged to update the PR on the next few months.

Taking the burden off of the reviewers

grigi commented on docs/CONTRIBUTING.md:
```diff
@@ -1,18 +1,238 @@
+
+The Klipper "reviewers" are:
+
+| Name                   | GitHub name       | Areas of interest |
+| ---------------------- | ----------------- | ----------------- |
+| Eric Callahan          | @Arksine          | Bed leveling, MCU flashing |
+| Dmitry Butyugin        | @dmbutyugin       | Input shaping, resonance testing, kinematics |
+| Kevin O'Connor         | @KevinOConnor     | Core motion system, Micro-controller code |
```
Commits are prefixed by module. Could we consider prefixing PR titles by affected modules as well?

How does these subsystems translate to areas of interest?

grigi reviewed (commented):
This document help clarifies quite a few of the unwritten rules, and this is good :+1:

KevinOConnor commented:
Thanks.  I made some updates to the contributing document based on feedback I've seen so far.

* I tried to streamline the "submit a PR" to "merge a PR" steps.  In particular, using Klipper Discourse is listed more as a hint then an official step.
* Add some info on the use of `git rebase` to fix merge conflicts.

@grigi 
> If we have been working on a PR, and once its deemed complete, how do we notify anyone that it's worth their time to actually pay attention?

Yeah.  Dmitry also mentioned this in a separate conversation.  Hopefully the new wording is a little more straight-forward.

> Could we consider prefixing PR titles by affected modules as well?

I'm not sure what you are suggesting.  I suspect the number of PRs is still small enough that reviewers will be able to find the ones they are interested in reviewing.

-Kevin

charlespick commented:
It's definitely a good thing to add more people to the process to get through the increasing activity around the project. However I will say there is one issue I see: I feel Klipper may miss out on some contributions because people fear they may not get a reviewer. It's tough to write code without even the likelihood someone will review it, much more it get merged. Although I don't have a solution either :(

phil65 commented:
Perhaps a pull request template could help to improve the general quality of PRs?
https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository

mental405 commented:
> It's definitely a good thing to add more people to the process to get through the increasing activity around the project. However I will say there is one issue I see: I feel Klipper may miss out on some contributions because people fear they may not get a reviewer. It's tough to write code without even the likelihood someone will review it, much more it get merged. Although I don't have a solution either :(

Not nearly as many contributions as the ones that don't even get a developer.

I don't feel like this should be seen as any sort of hinderance. The intent of this proposal is to get _more_ eyes on the submissions so that those that have merit can potentially be merged/tested sooner.  If a situation arises where a new feature or whatever is proposed and nobody is able to look at it, I am sure a developer could be poked with a large enough stick to give it at least a cursory examination.

charlespick commented:
@mental405 except the proposal explicitly says no poking developers and reviewers with sticks. I guess it stands to be seen how many people volunteer to review PRs. 

Fwiw I can review documentation prs

dmbutyugin commented:
Well, it was similar before, but with only a single person reviewing the PRs. Now with more people doing reviews, the overall throughput should get higher.

KevinOConnor commented:
> I feel Klipper may miss out on some contributions because people fear they may not get a reviewer. It's tough to write code without even the likelihood someone will review it, much more it get merged. Although I don't have a solution either

I understand.  I added the "don't ping the reviewers" blurb because I don't want the regular Klipper contributors to get "burned out".  I don't want to discourage new contributors - in particular some of those new contributors will turn into regular contributors.

I added the "who are the reviewers" section in hope that it would make the process more transparent.  I would not want that section to appear to be an "assignment of work", and definitely would not want volunteers feeling that they have some obligation to do work.

If there's a better way to word that, then let me know.

Cheers,
-Kevin

KevinOConnor commented:
> Perhaps a pull request template could help to improve the general quality of PRs?

Using templates, workflows, and automations was informally discussed earlier.  My experience is that the effort to setup and maintain these automations just adds more burden to a few people - a scalable solution really needs more people to be involved.

With respect to templates, the fear is that too many submitters treat them as an "annoying captcha" to be "mindlessly clicked through".  Others expressed their experience that they ultimately yield little value.

-Kevin

phil65 commented:
Ok, was just a thaught. I contributed extensively to the Open source project Kodi in the past and when we introduced a PR template there some years ago, it helped a bit. Just a link to the "Contributing" document would already be enough to at have a bit less work on maintainer side.

garethky commented:
Could we encourage contributors to solicit a review from other contributors who are not official reviewers? This makes for good practice, community building, many eyes etc. Not in place of an official review, but in preparation for an official review.

Should we outline a workflow so you get fewer commits to merge? At work we do squash merges so its always 1 commit. I know in a previous discussion @KevinOConnor mentioned that fewer commits meant that it would be easier to remove broken code from a release. This is the workflow I have cooked up to do that:

### PR Update Process
1. If needed, roll back the topic branch to the last commit that's part of your change. E.g. if you did a fast forward merge on GitHub you can remove it from the PR by going back to your last actual code commit.
    ```
    git checkout your-pr-topic-branch
    git reset --hard <SHA>
    ```
1. Change the topic branch upstream to master:
    ```
    git branch -u master
    ```
1. Go on github and pull in master from the kilpper origin
1. Then pull down those changes into your local master branch
    ```
    git checkout master
    git pull --rebase
    git checkout your-pr-topic-branch
    ```
1. Rebase your topic branch on top of masters commits:
    ```
    pull --rebase
    ```
1. Do whatever work needs doing! Use rebase interactive to collapse your commits.
    ```
    git commit -m "fixup"
    git rebase -i
    ```
1. When work is finished, switch upstream back to the remote:
    ```
    git branch -u origin/your-pr-topic-branch
    ```
1. Overwrite the remote PR branch with your prepared change using force push
    ```
    git push --force origin
    ```

KevinOConnor commented:
> Could we encourage contributors to solicit a review from other contributors who are not official reviewers?

Yeah - that would be a good idea.  It would definitely help to get more people to help with reviewing - it's not necessary to have any special privileges to perform a review.

> Should we outline a workflow so you get fewer commits to merge?

Generally it's not hard to squash a complete change into a single commit via the github web interface.  So, squashing down commits is usually not an issue.  For large changes, it's often necessary to break them up into separate commits (and we do want to commit them as separate).  When we do want separate commits, it helps a lot to have each individual commit follow the guidelines in the documentation (because, it's not easy to merge individual commits with changes to them).

So, for small changes or new modules, it's generally good to get a single commit, but for large changes, or changes that impact many different modules then we really want a series of commits.

I agree that getting used to git and `git rebase` can be a challenge.  Describing that process is a lot of content though, so ideally something that doesn't need to be hosted on the Klipper docs.  Other projects (eg, QEMU at https://qemu.readthedocs.io/en/latest/devel/submitting-a-patch.html ) have quite a bit of documentation, so maybe we can link to something useful online.

Thanks for the feedback,
-Kevin

mhier commented:
I like it. Two random thoughts (from a random guy :-D):

The document states:
> A reviewer may ask for clarification on how a user is to configure an option [...]

I would expect that information like this is part of the documentation, which should be included in the PR. Otherwise the reviewer knows how to configure the new module but nobody else ;-)

Then, if it helps, you could implement some kind of "pre-review" system, where some pre-reviewers just check the PR for some formal criteria which do not require much thinking or knowledge (e.g. tests are passing, commit messages are properly formatted etc.). This could reduce the work-load on the actual reviewers. Pre-reviewers would not need to be as much "qualified" as the real reviewers (e.g. more or less newcomers like me could do this job, while I wouldn't fit into the category to have "an ongoing history working with users [...]" etc.). This could be organised in a way that pre-reviewers cannot do real harm (the real reviewer would still check these points briefly, but likely he does not have to request changes for them). This is just an idea... And maybe it would be an opportunity for me to help :-)

ayufan commented:
I like this doc. It seems a step in a good direction.

I'm sometimes contributing to `esphome/esphome`. They implemented a similar reviewer system slightly more automated. Each component has a defined `CODEOWNER` that instructs a bot to ping the person in question automatically that contributed or is main reviewer for a given component.

Maybe some ideas from https://esphome.io/guides/contributing.html#contributing-to-esphomedocs could be adopted as well.

Klipper seems quite similar to this, where it should be possible to describe an owners of each particular `extras` or source. With the automation it could help direct to reviewers, but also do some static checks as well. As I assume that most of people including myself does forget about `signoff` line :)

eduncan911 commented:
I'll put my hat in as fully automating the entire processes with GitHub Actions.  Over written a number of my own actions, as well as modifying existing ones to fit my needs.

I read above where Kevin said automation and templates may be too much of a burden.  That is true, when they get in the users' way.

I tend to architect my solutions for enterprise companies in a way that there is automation red tape you have to do (must be signed, must be current commits, etc), but beyond that users are free to PR and write issues.

Templates for PRs and Issues tend to be overly restrictive and road blocks.  I recommend against templates, beyond some helpful advice at the beginning of the PR and Issues.

I also mentioned above where a BOT (I usually write them in github actions now) would be responsible for offloading the Reviewers burdens in maintaining the PR's state: **Reviewers should have the simple task of popping in, looking through a list of Needs Review, or Needs Follow-Up Review, and that's it.  No special comments, no special tagging, etc.  The Bot would handle all of that.

And finally, the bot would interact with the submitter mostly by auto tagging, auto commenting, kicking back a PR that isn't signed, etc.

Just ping me here if you'd like the help.  I'm also fast at code quality checks.

KevinOConnor commented:
I made some updates to the proposed workflow:
1. I added a section to make it clear that help with reviews is appreciated and that it isn't necessary to be a "listed reviewer" to do a review.
2. I changed the proposal to use github's existing "under review" and "review approved" flags instead of using labels to indicate the review status.

-Kevin

KevinOConnor commented:
Thanks for the feedback!

Some random thoughts..

> > A reviewer may ask for clarification on how a user is to configure an option [...]

> I would expect that information like this is part of the documentation, which should be included in the PR. Otherwise the reviewer knows how to configure the new module but nobody else ;-)

I agree it would be good to get detailed documentation on each change.  Alas, I think requiring documentation on "how to configure" every config option might be asking for too much.  (For example, a document describing "how to configure max_accel" could be pages of text.)  In the section cited above I was thinking more that a reviewer should be able to identify, at least in their own minds, how a user would use a config option.  And if they can't it should be clarified (at least as far as the reviewer can understand it in their own mind).

> Then, if it helps, you could implement some kind of "pre-review" system, where some pre-reviewers just check the PR for some formal criteria which do not require much thinking or knowledge

Yeah - I updated the doc to try to make it clear that we appreciate help with reviews and that one doesn't need to be a "listed reviewer" to help.

> I'm sometimes contributing to esphome/esphome. They implemented a similar reviewer system, but slightly more automated (and maybe more distributed). Each component has a defined CODEOWNER that instructs a bot to ping the person in question automatically that contributed or is main reviewer for a given component.

Thanks for the link.  I agree we could do something like that.  I think we might want to get experience with this workflow and see what automation helps after a few months of using it.

> I'll put my hat in as fully automating the entire processes with GitHub Actions.

FWIW, the most important part of the review to me is: 1) is the change defective, and 2) does the change provide a "high impact benefit"?  I don't know of any way to automate those checks.  If the above passes, we should be able to get through all the other stuff.  In the past I've seen that automation is sometimes treated as a "captcha" to be mindlessly clicked through (eg, create a signed-off-by line with a pseudonym to get past the bot). I fear that just makes more work for everyone.

That said, if we run into a common pattern of events with the new review system that can be automated, then I'm certainly open to using it.  The Klipper github repo has no shortage of github actions already.

Thanks again for the feedback!
-Kevin

KevinOConnor commented:
Thanks for the comments.  I have committed this change.

-Kevin
